go/src/socialapi/workers/countly/client/countly.go

```diff
@@ -2,75 +2,105 @@ package client
 
 import (
 	"encoding/json"
+	"errors"
+	"io/ioutil"
 	"net/http"
 	"net/url"
 	"strings"
+
+	"github.com/koding/logging"
+)
+
+var (
+	defaultDomainURL = "http://localhost:32768"
 )
 
+// Client holds required properties for communicating with a countly api.
 type Client struct {
-	defaultClient *http.Client
-	token         string
-	baseURL       string
+	httpClient *http.Client
+	token      string
+	baseURL    string
+	log        logging.Logger
 }
 
 // Option is a type for Client options to pass while creating client
 type Option func(*Client)
 
-var (
-	defaultDomainURL = ""
-)
-
 // New initializes the countly api client with given token and option parameters
 func New(apiKey string, opts ...Option) *Client {
 	if apiKey == "" {
-		panic("invalid api token")
+		panic("api token is required")
 	}
 
 	client := &Client{
-		defaultClient: &http.Client{},
-		token:         apiKey,
-		baseURL:       defaultDomainURL,
+		httpClient: http.DefaultClient,
+		token:      apiKey,
+		baseURL:    defaultDomainURL,
 	}
 
 	for _, option := range opts {
 		option(client)
 	}
 
+	if client.log == nil {
+		logging.NewCustom("countly client", false)
+	}
+
 	return client
 }
 
 // SetBaseURL sets the url option for Client struct
 func SetBaseURL(url string) Option {
 	return func(c *Client) {
-		c.baseURL = url
+		c.baseURL = strings.TrimSuffix(url, "/")
 	}
 }
 
-var defaultClient = &http.Client{}
+// SetClient sets the client  option for Client struct
+func SetClient(cl *http.Client) Option {
+	return func(c *Client) {
+		c.httpClient = cl
+	}
+}
+
+// SetLogger sets the logger option for Client
+func SetLogger(log logging.Logger) Option {
+	return func(c *Client) {
+		c.log = log
+	}
+}
 
 func (c *Client) do(method, path string, values url.Values, v interface{}) error {
 	u := c.createURL(path, values)
+	c.log.Debug("url for request %s", u.String())
+
 	req, err := http.NewRequest(method, u.String(), nil)
 	if err != nil {
 		return err
 	}
 
-	resp, err := c.defaultClient.Do(req)
+	resp, err := c.httpClient.Do(req)
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 
-	return json.NewDecoder(resp.Body).Decode(&v)
-}
+	b, err := ioutil.ReadAll(resp.Body)
```

Comment:
> What about setting some body limit?

even tho i am not sure at this point how much data we are going to fetch from countly api, 4<< 20 is a big number that we could assume as "safe". will add "LimitReader" ( i have only seen couple hundreds bytes of data received )


> Is it possible to have nil v? (e.g. no payload)

didnt check their whole api yet, hence it is hidden under our worker not public/standalone